src/Raven.Server/Documents/Handlers/Debugging/DocumentDebugHandler.cs

```diff
@@ -54,5 +57,89 @@ public async Task HugeDocuments()
+            var corrupted = new List<string>();
+            long? lastEtag = null;
+            
+            using (ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+            {
+                await using (var writer = new AsyncBlittableJsonTextWriter(context, Stream.Null))
+                using (context.OpenReadTransaction())
+                {
+                    foreach (var doc in IterateDocumentsAndRevisionsByEtag(context, startEtag))
+                    {
+                        HttpContext.RequestAborted.ThrowIfCancellationRequested();
+                        
+                        using (doc)
+                        {
+                            try
+                            {
+                                writer.WriteString(doc.Id);
+                            }
+                            catch (Exception)
+                            {
+                                corrupted.Add($"Id: '{doc.Id}', LowerId: '{doc.LowerId}', ChangeVector: '{doc.ChangeVector}', Etag: '{doc.Etag}', Flags: '{doc.Flags}'");
+
+                                if (corruptedCount-- <= 0)
+                                {
+                                    lastEtag = doc.Etag;
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            
+                await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream()))
+                {
+                    writer.WriteStartObject();
+                    writer.WriteArray("CorruptedDocuments", corrupted);
+                    writer.WriteComma();
+                    if (lastEtag.HasValue)
+                    {
+                        writer.WritePropertyName("LastEtag");
+                        writer.WriteInteger(lastEtag.Value);
+                    }
+                    else
+                    {
+                        writer.WritePropertyName("AllScanned");
+                        writer.WriteBool(true);
+                    }
+                    writer.WriteEndObject();
+                }
+            }
+        }
+        
+        public IEnumerable<Document> IterateDocumentsAndRevisionsByEtag(DocumentsOperationContext context, long startEtag)
+        {
+            const DocumentFields documentFields = DocumentFields.Id | DocumentFields.LowerId | DocumentFields.ChangeVector;
+            var documents = Database.DocumentsStorage.GetDocumentsFrom(context, startEtag, 0, long.MaxValue, documentFields);
+            var revisions = Database.DocumentsStorage.RevisionsStorage.GetRevisionsFrom(context, startEtag, long.MaxValue, documentFields);
+            
+            using var documentsEnumerator = documents.GetEnumerator(); 
+            using var revisionsEnumerator = revisions.GetEnumerator(); 
+            
+            var enumerators = new List<IEnumerator<Document>>();
```

Comment:
use the `MergedEnumerator`?

Reply:
Are you referring to MergedReplicationBatchEnumerator?  
It is too tightly coupled with replication.  
I thought it would be much easier and cleaner to just mimic its behavior, so that’s what I did.  

I could refactor it to have a base class that I can use, but such a refactor would take a long time for the PR to get approved.  
I can open an issue for that if you’d like, but anyway this scan is only for fixing the bug.  

Alternatively, I can use MergedReplicationBatchEnumerator and push in all the stats it needs just to satisfy it.  
I don’t really like this approach, but if that’s what you prefer, I’ll do it.

Reply:
I believe it is safe enough simply to copy it to v5